cpp/src/arrow/types/struct-test.cc

```diff
@@ -52,4 +62,133 @@ TEST(TestStructType, Basics) {
   // TODO(wesm): out of bounds for field(...)
 }
 
-}  // namespace arrow
+// .............................................................................
+// Struct test
+class TestStructBuilder : public TestBuilder {
+ public:
+  void SetUp() {
+    TestBuilder::SetUp();
+
+    auto value_type = TypePtr(new Int32Type());
+    auto char_type = TypePtr(new Int8Type());
+    auto list_type = TypePtr(new ListType(char_type));
+
+    std::vector<TypePtr> types = {list_type, value_type};
+    std::vector<FieldPtr> fields;
+    fields.push_back(FieldPtr(new Field("list", list_type)));
+    fields.push_back(FieldPtr(new Field("int", value_type)));
+
+    type_ = TypePtr(new StructType(fields));
+    value_fields_ = fields;
+
+    std::shared_ptr<ArrayBuilder> tmp;
+    ASSERT_OK(MakeStructBuilder(pool_, type_, fields, &tmp));
+
+    builder_ = std::dynamic_pointer_cast<StructBuilder>(tmp);
+  }
+
+  void Done() {
+    result_ = std::dynamic_pointer_cast<StructArray>(builder_->Finish());
+  }
+
+ protected:
+  std::vector<FieldPtr> value_fields_;
+  TypePtr type_;
+
+  std::shared_ptr<StructBuilder> builder_;
+  std::shared_ptr<StructArray> result_;
+};
+
+TEST_F(TestStructBuilder, TestAppendNull) {
+  ASSERT_OK(builder_->AppendNull());
+  ASSERT_OK(builder_->AppendNull());
+  ASSERT_EQ(2, builder_->value_builder().size());
+
+  Done();
+
+  ASSERT_EQ(2, result_->values().size());
+  ASSERT_EQ(2, result_->length());
+  ASSERT_TRUE(result_->IsNull(0));
+  ASSERT_TRUE(result_->IsNull(1));
+
+
+  auto list_char = static_cast<ListArray*>(result_->values(0).get());
+  auto chars = static_cast<Int8Array*>(list_char->values().get());
+  auto int32 = static_cast<Int32Array*>(result_->values(1).get());
+  ASSERT_EQ(0, list_char->length());
+  ASSERT_EQ(0, chars->length());
+  ASSERT_EQ(0, int32->length());
+
+  ASSERT_EQ(Type::LIST, list_char->type_enum());
+  ASSERT_EQ(Type::INT8, list_char->values()->type_enum());
+  ASSERT_EQ(Type::INT32, int32->type_enum());
+}
+
+TEST_F(TestStructBuilder, TestBasics) {
+  vector<int32_t> int_values = {1, 2, 3, 4};
+  vector<char> list_values = {'j', 'o', 'e', 'b', 'o', 'b', 'm', 'a', 'r', 'k'};
+  vector<int> list_lengths = {3, 0, 3, 4};
+  vector<int> list_offsets = {0, 3, 3, 6, 10};
+  vector<uint8_t> list_is_not_null = {1, 0, 1, 1};
+  vector<uint8_t> struct_is_not_null = {1, 1, 1, 1};
+
+  ListBuilder* list_vb = static_cast<ListBuilder*>(
+      builder_->value_builder().at(0).get());
+  Int8Builder* char_vb = static_cast<Int8Builder*>(
+      list_vb->value_builder().get());
+  Int32Builder* int_vb = static_cast<Int32Builder*>(
+      builder_->value_builder().at(1).get());
```

Comment:
for the StructBuilder, perhaps these methods should be `builder_->field(i)` to be more clear as noted above